### PR TITLE
Making json string to JsonString conversion explicit by removing JsonString From<String>

### PR DESCRIPTION
**Pull request by willemolding**
Re-attempt at #810

Currently using the hdk::call function in zome code requires manually parsing the JsonString since it is impossible to serialize into a ZomeApiResult. This is the first change to JsonString required to make this possible.

### CODE REVIEW

**Comment by willemolding** [comment]
Note this will probably be a breaking change for DNA developers if they are using JsonString

File: conductor_api/wasm-test/src/lib.rs [hunk 1]
```diff
@@ -109,7 +109,7 @@ pub extern "C" fn debug_stacked_hello(
     let fish = mem_stack.write_json(TestStruct {
         value: "fish".to_string(),
     });
-    hdk_debug(&mut mem_stack, &JsonString::from("disruptive debug log"));
+    hdk_debug(&mut mem_stack, &JsonString::from(RawString::from(RawString::from("disruptive debug log"))));
```
Comment (thedavidmeister) [inline conductor_api/wasm-test/src/lib.rs]: why the double `RawString`?

File: conductor_api/test-bridge-caller/src/lib.rs [hunk 2]
```diff
@@ -12,7 +12,7 @@ use hdk::holochain_core_types::{
 };
 
 fn handle_call_bridge() -> JsonString {
-    hdk::call("test-callee", "greeter", Address::from("token"), "hello", JsonString::from("{}")).unwrap()
+    hdk::call("test-callee", "greeter", Address::from("token"), "hello", JsonString::from_json("{}")).unwrap()
```
Comment (thedavidmeister) [inline conductor_api/test-bridge-caller/src/lib.rs]: `JsonString::empty_object()`?

File: core_types/src/json.rs [hunk 3]
```diff
@@ -39,14 +39,16 @@ impl JsonString {
...
+    // Creates a JsonString from stringified json
+    // replaces From<String> for JsonString and requires conversions to be explicit
+    // This is because string types must be handled differently depending on if
+    // they are strinfigied JSON or JSON strings
+    pub fn from_json(s: &str) -> JsonString {
         let cleaned = s
             // remove whitespace from both ends
             .trim()
-            // remove null characters from both ends
+            // remove null characters from both endsi
```
Comment (thedavidmeister) [inline core_types/src/json.rs]: typo

File: core_types/src/json.rs [hunk 4]
```diff
@@ -145,38 +147,47 @@ pub trait JsonError {}
...
+    }
+}
+
+impl<E: Into<JsonString>> From<Result<String, E>> for JsonString {
+    fn from(result: Result<String, E>) -> JsonString {
+        result_to_json_string(result.map(|v| RawString::from(v)))
+    }
+}
+
+impl From<Result<String, String>> for JsonString {
```
Comment (thedavidmeister) [inline core_types/src/json.rs]: @lucksus lookit ^^ better late than never?

File: nodejs_waiter/src/waiter.rs [hunk 5]
```diff
@@ -417,7 +417,7 @@ mod tests {
     }
 
     fn zf_response(call: ZomeFnCall) -> ExecuteZomeFnResponse {
-        ExecuteZomeFnResponse::new(call, Ok(JsonString::from("")))
+        ExecuteZomeFnResponse::new(call, Ok(JsonString::from_json("{}")))
```
Comment (thedavidmeister) [inline nodejs_waiter/src/waiter.rs]: `JsonString::empty_object()`

File: core_types/src/json.rs (hunk 4 above)
Comment (lucksus) [reply core_types/src/json.rs]: 👏

File: core_types/src/json.rs [hunk 6]
```diff
@@ -145,38 +147,47 @@ pub trait JsonError {}
...
+    let inner_json: JsonString = match result {
+        Ok(inner) => inner.into(),
+        Err(inner) => inner.into(),
+    };
+    let inner_string = String::from(inner_json);
+    JsonString::from_json(&format!(
+        "{{\"{}\":{}}}",
+        if is_ok { "Ok" } else { "Err" },
+        inner_string
+    ))
```
Comment (maackle) [inline core_types/src/json.rs]: This looks correct but doesn't Result serialize to exactly this anyway? `serde_json::to_string(&result)`?

Comment (thedavidmeister) [reply core_types/src/json.rs]: i think that's the point right? to extend the types/traits so that they "magically" track default serde behaviour

**Pull request merged**